src/include/distributed/multi_physical_planner.h

```diff
@@ -169,6 +169,9 @@ typedef struct Task
 	uint64 shardId;                /* only applies to shard fetch tasks */
 	TaskExecution *taskExecution;  /* used by task tracker executor */
 	bool upsertQuery;              /* only applies to modify tasks */
+
+	bool insertSelectQuery;
```

Comment:
What if we add an explicit routerExecutable flag to the job? That will survive the test of time.

I could imagine a routerExecutable job that is not an INSERT/SELECT. This particular locking behaviour is quite specific to INSERT/SELECT tasks (or at least writes with a co-located subquery).